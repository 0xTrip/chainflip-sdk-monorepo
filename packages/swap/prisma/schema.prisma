// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

model SwapIntent {
  id             BigInt   @id @default(autoincrement())
  uuid           String   @unique @default(uuid())
  ingressAsset   String
  ingressAddress String
  egressAsset    String
  egressAddress  String
  active         Boolean  @default(true)
  createdAt      DateTime @default(now())
  swaps          Swap[]

  @@index([ingressAddress])
  @@schema("public")
}

model Swap {
  id                BigInt     @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId          BigInt     @unique
  ingressAmount     Decimal    @db.Decimal(30, 0)
  ingressReceivedAt DateTime
  swapExecutedAt    DateTime?
  egressCompleteAt  DateTime?
  swapIntent        SwapIntent @relation(fields: [swapIntentId], references: [id])
  swapIntentId      BigInt
  egress            Egress?
  createdAt         DateTime   @default(now())
  updatedAt         DateTime   @default(now()) @updatedAt

  @@schema("public")
}

enum Network {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  network     Network
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, network])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  network  Network
  egresses Egress[]

  @@unique([nativeId, network])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String  
  createdAt     DateTime           @default(now())

  @@schema("public")
}
