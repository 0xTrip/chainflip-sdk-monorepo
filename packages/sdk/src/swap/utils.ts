export const unreachable = (x: never, message: string): never => {
  throw new Error(`${message}: ${x}`);
};
