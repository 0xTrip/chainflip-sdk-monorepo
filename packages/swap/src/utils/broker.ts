import { u8aToHex } from '@polkadot/util';
import { decodeAddress } from '@polkadot/util-crypto';
import { z } from 'zod';
import { supportedAsset, SupportedAsset } from '@/shared/assets';
import {
  bareHexString,
  btcString,
  hexString,
  numericString,
} from '@/shared/parsers';
import { memoize } from './function';
import RpcClient from './RpcClient';
import { transformAsset } from './string';

type NewSwapIntent = {
  ingressAsset: SupportedAsset;
  egressAsset: SupportedAsset;
  egressAddress: string;
};

const address = z.union([
  hexString,
  bareHexString.transform((v) => `0x${v}`),
  btcString,
]);

const requestValidators = {
  newSwapIngressAddress: z
    .tuple([
      supportedAsset.transform(transformAsset),
      supportedAsset.transform(transformAsset),
      numericString.or(hexString),
      z.number(),
    ])
    .transform(([a, b, c, d]) => [a, b, c, d]),
};

const responseValidators = {
  newSwapIngressAddress: address,
};

const initializeClient = memoize(async () => {
  const rpcClient = await new RpcClient(
    process.env.RPC_RELAYER_WSS_URL as string,
    requestValidators,
    responseValidators,
    'relayer',
  ).connect();

  return rpcClient;
});

export const submitSwapToBroker = async (
  swapIntent: NewSwapIntent,
): Promise<string> => {
  const { ingressAsset, egressAsset, egressAddress } = swapIntent;
  const client = await initializeClient();
  const ingressAddress = await client.sendRequest(
    'newSwapIngressAddress',
    ingressAsset,
    egressAsset,
    egressAsset === 'DOT' // btc not yet implemented on broker side
      ? u8aToHex(decodeAddress(egressAddress))
      : egressAddress,
    0,
  );

  return ingressAddress;
};
