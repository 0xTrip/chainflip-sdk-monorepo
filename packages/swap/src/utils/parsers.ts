import { z } from 'zod';

export const thirdPartySwapSchema = z.object({
  uuid: z.string(),
  routeResponse: z
    .object({
      protocol: z.enum(['Lifi', 'Squid']),
    })
    .passthrough(), // pass through routeResponse objects until we are fully certain of its shape
});
