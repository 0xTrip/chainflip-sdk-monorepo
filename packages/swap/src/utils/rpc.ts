import { UncheckedAssetAndChain } from '@/shared/enums';
import {
  getIngressEgressEnvironment,
  getSwappingEnvironment,
} from '@/shared/rpc';
import {
  readAssetValue,
  validateSwapAmount as validateAmount,
} from '@/shared/rpc/utils';
import { memoize } from './function';

const cachedGetSwappingEnvironment = memoize(getSwappingEnvironment, 60_000);
const cachedGetIngressEgressEnvironment = memoize(
  getIngressEgressEnvironment,
  6_000,
);

type Result = { success: true } | { success: false; reason: string };

export const validateSwapAmount = async (
  asset: UncheckedAssetAndChain,
  amount: bigint,
): Promise<Result> => {
  const swapEnv = await cachedGetSwappingEnvironment({
    rpcUrl: process.env.RPC_NODE_HTTP_URL as string,
  });

  return validateAmount(swapEnv, asset, amount);
};

export const getIngressFee = async (
  asset: UncheckedAssetAndChain,
): Promise<bigint> => {
  const ingressEgressEnv = await cachedGetIngressEgressEnvironment({
    rpcUrl: process.env.RPC_NODE_HTTP_URL as string,
  });

  return ingressEgressEnv.ingressFees
    ? readAssetValue(ingressEgressEnv.ingressFees, asset)
    : 0n;
};

export const getEgressFee = async (
  asset: UncheckedAssetAndChain,
): Promise<bigint> => {
  const ingressEgressEnv = await cachedGetIngressEgressEnvironment({
    rpcUrl: process.env.RPC_NODE_HTTP_URL as string,
  });

  return ingressEgressEnv.egressFees
    ? readAssetValue(ingressEgressEnv.egressFees, asset)
    : 0n;
};
