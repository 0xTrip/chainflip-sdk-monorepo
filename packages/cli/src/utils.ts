import { createInterface } from 'node:readline/promises';
import yargs from 'yargs/yargs';
import { Assets, ChainflipNetworks } from '@/shared/enums';
import { chainflipNetwork } from '@/shared/parsers';
import { ChainflipNetwork } from './enums';

export const askForPrivateKey = async () => {
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  try {
    return await rl.question("Please enter your wallet's private key: ");
  } finally {
    rl.close();
  }
};

type GetEthNetworkOptions =
  | { chainflipNetwork: 'localnet'; ethNetwork: string }
  | { chainflipNetwork: ChainflipNetwork };

export function getEthNetwork(opts: GetEthNetworkOptions) {
  if (opts.chainflipNetwork === 'localnet') return opts.ethNetwork;
  if (opts.chainflipNetwork === ChainflipNetworks.mainnet) return 'mainnet';
  return 'goerli';
}
const networks = [...Object.values(chainflipNetwork.enum), 'localnet'];
export const parseArgs = (args: string[]) =>
  yargs(args)
    .scriptName('chainflip-cli')
    .usage('$0 <cmd> [args]')
    .command('swap', '', (y) => {
      y.option('src-asset', {
        choices: Object.values(Assets),
        // demandOption: true,
        describe: 'The asset to swap from',
      })
        .option('dest-asset', {
          choices: Object.values(Assets),
          demandOption: true,
          describe: 'The asset to swap to',
        })
        .option('chainflip-network', {
          choices: networks,
          describe: 'The Chainflip network to execute the swap on',
          default: ChainflipNetworks.sisyphos,
        })
        .option('amount', {
          type: 'string',
          demandOption: true,
          describe: 'The amount to swap',
        })
        .option('dest-address', {
          type: 'string',
          demandOption: true,
          describe: 'The address to send the swapped assets to',
        })
        .option('wallet-private-key', {
          type: 'string',
          describe: 'The private key of the wallet to use',
        })
        .option('src-token-contract-address', {
          type: 'string',
          describe:
            'The contract address of the token to swap from when `chainflip-network` is `localnet`',
        })
        .option('vault-contract-address', {
          type: 'string',
          describe:
            'The contract address of the vault when `chainflip-network` is `localnet`',
        })
        .option('eth-network', {
          type: 'string',
          describe:
            'The eth network URL to use when `chainflip-network` is `localnet`',
        });
    })
    .command('fund-state-chain-account', '', (y) => {
      y.option('account-id', {
        type: 'string',
        demandOption: true,
        describe: 'The account ID for the validator to be funded',
      })
        .option('chainflip-network', {
          choices: networks,
          describe: 'The Chainflip network to execute the swap on',
          default: ChainflipNetworks.sisyphos,
        })
        .option('amount', {
          type: 'string',
          demandOption: true,
          describe: 'The amount in Flipperino to fund',
        })
        .option('wallet-private-key', {
          type: 'string',
          describe: 'The private key of the wallet to use',
        })
        .option('state-chain-manager-contract-address', {
          type: 'string',
          describe:
            'The contract address of the state chain manager when `chainflip-network` is `localnet`',
        })
        .option('flip-token-contract-address', {
          type: 'string',
          describe:
            'The contract address for the FLIP token when `chainflip-network` is `localnet`',
        })
        .option('eth-network', {
          type: 'string',
          describe:
            'The eth network URL to use when `chainflip-network` is `localnet`',
        });
    })
    .help()
    .parseSync();
