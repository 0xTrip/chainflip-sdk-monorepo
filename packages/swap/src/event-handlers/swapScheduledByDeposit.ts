// Set the column in the DB to the block timestamp and the deposit amount.
import assert from 'assert';
import { z } from 'zod';
import { unsignedInteger } from '@/shared/parsers';
import { chainAddress } from './common';
import logger from '../utils/logger';
import type { EventHandlerArgs } from '.';

const eventArgs = z.object({
  depositAddress: chainAddress,
  swapId: unsignedInteger,
  depositAmount: unsignedInteger,
});

export type SwapScheduledByDepositEvent = z.input<typeof eventArgs>;

export default async function swapScheduledByDeposit({
  prisma,
  block,
  event,
}: EventHandlerArgs): Promise<void> {
  logger.warn('SwapScheduledByDeposit handler is here');
  try {
    // get necessary params
    const {
      depositAddress: { address: depositAddress },
      swapId,
      depositAmount,
    } = eventArgs.parse(event.args);

    // retrieve SwapDepositChannel for later processing
    const swapRequests = await prisma.swapDepositChannel.findMany({
      where: { depositAddress, active: true },
    });

    if (swapRequests.length === 0) {
      logger.info(
        `SwapScheduledByDeposit: SwapDepositChannel not found for depositAddress ${depositAddress}`,
      );
      return;
    }

    assert(
      swapRequests.length === 1,
      `SwapScheduledByDeposit: too many active swap intents found for depositAddress ${depositAddress}`,
    );

    // Create a new swap object
    await prisma.swap.create({
      data: {
        depositAmount: depositAmount.toString(),
        nativeId: swapId,
        depositReceivedAt: new Date(block.timestamp),
        swapDepositChannelId: swapRequests[0].id,
      },
    });
  } catch (error) {
    logger.customError(
      'error in "SwapScheduledByDeposit" handler',
      { alertCode: 'EventHandlerError' },
      { error, handler: 'SwapScheduledByDeposit' },
    );
    throw error;
  }
}
