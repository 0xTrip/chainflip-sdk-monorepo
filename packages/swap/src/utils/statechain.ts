import { z } from 'zod';
import { SupportedAsset, supportedAsset } from '@/shared/assets';
import { numericString } from '@/shared/parsers';
import { memoize } from './function';
import RpcClient from './RpcClient';
import { transformAsset } from './string';
import { QuoteResponse, QuoteQueryParams } from '../schemas';

const requestValidators = {
  swap_rate: z.tuple([
    supportedAsset.transform(transformAsset),
    supportedAsset.transform(transformAsset),
    numericString,
  ]),
};

// parse hex encoding or decimal encoding into decimal encoding
const assetAmount = z.string().transform((v) => BigInt(v).toString());

const responseValidators = {
  swap_rate: z.object({
    // TODO: simplify when we know how Rust `Option` is encoded
    intermediary: assetAmount.optional().nullable(),
    output: assetAmount,
  }),
};

const initializeClient = memoize(async () => {
  const rpcClient = await new RpcClient(
    process.env.RPC_NODE_WSS_URL as string,
    requestValidators,
    responseValidators,
    'cf',
  ).connect();

  return rpcClient;
});

const getSwapAmount = async (
  depositAsset: SupportedAsset,
  destinationAsset: SupportedAsset,
  amount: string,
): Promise<z.infer<(typeof responseValidators)['swap_rate']>> => {
  const client = await initializeClient();

  return client.sendRequest(
    'swap_rate',
    depositAsset,
    destinationAsset,
    amount,
  );
};

export const getBrokerQuote = async (
  { depositAsset, destinationAsset, amount }: QuoteQueryParams,
  id: string,
): Promise<QuoteResponse> => {
  const { intermediary, output } = await getSwapAmount(
    depositAsset,
    destinationAsset,
    amount,
  );

  return {
    id,
    intermediateAmount: intermediary ?? undefined,
    egressAmount: output,
  };
};
