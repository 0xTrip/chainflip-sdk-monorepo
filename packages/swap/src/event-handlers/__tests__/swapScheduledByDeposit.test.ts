import { swapScheduledByDepositMock } from './utils';
import prisma, { SwapDepositChannel } from '../../client';
import swapScheduledByDeposit from '../swapScheduledByDeposit';

const ETH_ADDRESS = '0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2';
const DOT_ADDRESS = '5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX';

type SwapData = Parameters<
  (typeof prisma)['swapDepositChannel']['create']
>[0]['data'];
const createSwapRequest = (
  data: Partial<SwapData> = {},
): Promise<SwapDepositChannel> =>
  prisma.swapDepositChannel.create({
    data: {
      depositAsset: 'ETH',
      destinationAsset: 'DOT',
      depositAddress: ETH_ADDRESS,
      destinationAddress: DOT_ADDRESS,
      expectedDepositAmount: '10000000000',
      ...data,
    },
  });

describe(swapScheduledByDeposit, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  it('stores a new swap', async () => {
    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createSwapRequest();

    // create a swap after receiving the event
    await prisma.$transaction(async (client) => {
      await swapScheduledByDeposit({
        prisma: client,
        block: swapScheduledByDepositMock.block as any,
        event: swapScheduledByDepositMock.eventContext.event as any,
      });
    });

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
    });

    expect(swap.depositAmount.toString()).toEqual(
      swapScheduledByDepositMock.eventContext.event.args.depositAmount,
    );
    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
    });
  });
});
