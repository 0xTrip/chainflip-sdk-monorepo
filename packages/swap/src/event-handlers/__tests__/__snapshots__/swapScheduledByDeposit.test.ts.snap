// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapScheduledByDeposit stores a new swap 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "222222222222222222",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "10-0",
  "egressCompletedAt": null,
  "egressCompletedBlockIndex": null,
  "id": Any<BigInt>,
  "nativeId": 9876545n,
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "updatedAt": Any<Date>,
}
`;
